ibc/module.go

```diff
@@ -69,91 +63,67 @@ func (m *ibcModule) Start() error {
 		m.logger.Info().Msg("🚫 IBC module disabled 🚫")
 		return nil
 	}
-	m.logger.Info().Msg("🪐 starting IBC module 🪐")
-	// TODO: start the host logic
+	m.logger.Info().Msg("✅ Starting IBC Module ✅")
 	return nil
 }
 
 func (m *ibcModule) Stop() error {
+	m.logger.Info().Msg("🛑 Stopping IBC Module 🛑")
 	return nil
 }
 
-func (m *ibcModule) GetHost() modules.IBCHost {
-	return m.host
-}
-
 func (m *ibcModule) GetModuleName() string {
 	return modules.IBCModuleName
 }
 
-// HandleMessage accepts a generic IBC message and routes it to the utility mempool
-func (m *ibcModule) HandleMessage(message *anypb.Any) error {
-	m.m.Lock()
-	defer m.m.Unlock()
-
-	// Check the message is actually a valid IBC message
-	msg, err := codec.GetCodec().FromAny(message)
-	if err != nil {
-		return err
-	}
-	ibcMessage, ok := msg.(*ibcTypes.IBCMessage)
-	if !ok {
-		return fmt.Errorf("failed to cast message to IBCMessage")
-	}
-	if err := ibcMessage.ValidateBasic(); err != nil {
-		return err
-	}
-
-	// Convert IBC message to a utility Transaction
-	tx, err := ibcTypes.ConvertIBCMessageToTx(ibcMessage)
-	if err != nil {
-		return err
-	}
-
-	// Sign the transaction
-	pkBz, err := hex.DecodeString(m.cfg.PrivateKey)
-	if err != nil {
-		return err
-	}
-	pk, err := crypto.NewPrivateKeyFromBytes(pkBz)
+func (m *ibcModule) HandleEvent(event *anypb.Any) error {
+	evt, err := codec.GetCodec().FromAny(event)
 	if err != nil {
 		return err
 	}
-	signableBz, err := tx.SignableBytes()
-	if err != nil {
-		return err
-	}
-	signature, err := pk.Sign(signableBz)
-	if err != nil {
-		return err
-	}
-	tx.Signature = &coreTypes.Signature{
-		Signature: signature,
-		PublicKey: pk.PublicKey().Bytes(),
-	}
 
-	// Marshall the Transaction and send it to the utility module
-	txBz, err := codec.GetCodec().Marshal(tx)
-	if err != nil {
-		return err
-	}
-	if err := m.GetBus().GetUtilityModule().HandleTransaction(txBz); err != nil {
-		return err
+	switch event.MessageName() {
+	case messaging.ConsensusNewHeightEventType:
+		consensusNewHeightEvent, ok := evt.(*messaging.ConsensusNewHeightEvent)
+		if !ok {
+			return fmt.Errorf("failed to cast event to ConsensusNewHeightEvent")
+		}
+		newHeight := consensusNewHeightEvent.GetHeight()
+		// check if host was created exit if not
+		if m.host == nil {
+			break
+		}
+		// Flush all caches to disk for last height
+		bsc := m.GetBus().GetBulkStoreCacher()
+		if err := bsc.FlushAllEntries(); err != nil {
+			return err
+		}
+		// Prune old cache entries
+		if newHeight <= m.cfg.Host.BulkStoreCacher.MaxHeightStored {
+			break
+		}
+		pruneHeight := newHeight - m.cfg.Host.BulkStoreCacher.MaxHeightStored
+		if err := bsc.PruneCaches(pruneHeight); err != nil {
```

Comment:
Isn't this always going to prune small heights?

E.g. If new height is `102` and max height stores is `100`, we'll be running height 2?

I feel like you're attempt is having a circular buffer in the cache but I'm not convinced that's what it's actually doing right now.

Reply:
So what this does is we are at height 100, we have max stored height of 5, so we have stored 95-100 heights in our cache. when 101 comes in we prune 95 and now store 96-101. The cache doesnt need to be very large just enough to be able to replay if we feel the tx didnt propagate. max height stored is customisable but defaults to 5